pkg/processor/controlcommunication/types.go

```diff
@@ -132,13 +184,21 @@ func (acmb *AbstractControlMessageBroker) ReadControlMessage(reader *bufio.Reade
 }
 
 func (acmb *AbstractControlMessageBroker) SendToConsumers(message *ControlMessage) error {
-	acmb.channelLock.Lock()
-	defer acmb.channelLock.Unlock()
 
 	for _, consumer := range acmb.Consumers {
 		if consumer.GetKind() == message.Kind {
-			if err := consumer.Send(message); err != nil {
-				return errors.Wrap(err, "Failed to send message to consumer")
+			switch message.Kind {
+			case DrainDoneMessageKind:
+				if err := consumer.BroadcastAndCloseSubscriptions(message); err != nil {
+					return errors.Wrap(err, "Failed to send message to consumer")
+				}
+			case StreamMessageAckKind:
+				if err := consumer.Broadcast(message); err != nil {
+					return errors.Wrap(err, "Failed to broadcast message to consumer")
+				}
+			default:
+				// if message kind is not known, then just exit from the loop
```

Comment:
I think it's better to return an error.
It will be handled in the caller (for now, `controlOutputHandler` just logs the error and ignores it

```golang
return errors.Wrap(err, 
    "Received unknown control message kind",
    "kind", message.Kind)
```